From 80bb5463b641ab97c7846c53e33ad9bb3e919e9d Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Mon, 21 Apr 2014 23:26:49 +0800
Subject: [PATCH] mute selected font not match warning

---
 libass/ass_fontconfig.c | 15 ---------------
 1 file changed, 15 deletions(-)

diff --git a/libass/ass_fontconfig.c b/libass/ass_fontconfig.c
index b8ad9ec..84958d6 100644
--- a/libass/ass_fontconfig.c
+++ b/libass/ass_fontconfig.c
@@ -255,21 +255,6 @@ static char *select_font(ASS_Library *library, FCInstance *priv,
     if (result != FcResultMatch)
         r_fullname = NULL;
 
-    if (!treat_family_as_pattern &&
-        !(r_family && strcasecmp((const char *) r_family, family) == 0) &&
-        !(r_fullname && strcasecmp((const char *) r_fullname, family) == 0)) {
-        char *fallback = (char *) (r_fullname ? r_fullname : r_family);
-        if (code) {
-            ass_msg(library, MSGL_WARN,
-                    "fontconfig: cannot find glyph U+%04X in font '%s', falling back to '%s'",
-                    (unsigned int)code, family, fallback);
-        } else {
-            ass_msg(library, MSGL_WARN,
-                    "fontconfig: cannot find font '%s', falling back to '%s'",
-                    family, fallback);
-        }
-    }
-
     result = FcPatternGetString(rpat, FC_STYLE, 0, &r_style);
     if (result != FcResultMatch)
         r_style = NULL;
-- 
1.9.2

