


diff -cibwr libX11-1.5.0/modules/im/ximcp/imDefLkup.c libX11-1.5.0-CI-mods/modules/im/ximcp/imDefLkup.c
*** libX11-1.5.0/modules/im/ximcp/imDefLkup.c	Sat Jun  2 15:37:18 2012
--- libX11-1.5.0-CI-mods/modules/im/ximcp/imDefLkup.c	Tue Oct 16 17:28:07 2012
***************
*** 370,375 ****
--- 370,390 ----
  
      (void)_XimRespSyncReply(ic, buf_s[0]);
  
+     /* FIXME */
+     if( KeyPress <= ev.type && ev.type <= ButtonRelease) {
+       char *cp = "unknown ";
+       if(ev.type == KeyPress)
+ 	cp = "k press  ";
+       if(ev.type == KeyRelease)
+ 	cp = "k release";
+       if(ev.type == ButtonPress)
+ 	cp = "b press  ";
+       if(ev.type == ButtonRelease)
+ 	cp = "b release";
+       
+       fprintf(stderr,"%s,%d: putback %s ev.xkey.time=%ld\n", __FILE__, __LINE__, cp, ev.xkey.time);
+     }
+     
      XPutBackEvent(d, &ev);
  
      return True;
***************
*** 713,718 ****
--- 728,738 ----
      ev.keycode = 0;
      ev.state = 0;
  
+     /* FIXME */
+     
+     fprintf(stderr,"%s,%d: putback k press   ev.time=%lu (bogus?) set to 0 after this.\n", __FILE__, __LINE__, ev.time);
+     ev.time = 0L;
+ 
      XPutBackEvent(im->core.display, (XEvent *)&ev);
  
      return True;
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imDefLkup.c~
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imDefLkup.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imDefLkup.o
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imDispch.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imDispch.o
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imEvToWire.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imEvToWire.o
diff -cibwr libX11-1.5.0/modules/im/ximcp/imExten.c libX11-1.5.0-CI-mods/modules/im/ximcp/imExten.c
*** libX11-1.5.0/modules/im/ximcp/imExten.c	Sat Jun  2 15:37:18 2012
--- libX11-1.5.0-CI-mods/modules/im/ximcp/imExten.c	Thu Oct 11 18:29:09 2012
***************
*** 156,161 ****
--- 156,164 ----
      kev->state		= buf_s[3];		/* state */
      kev->time		= buf_l[2];		/* time */
  
+     /*FIXME*/
+     fprintf(stderr,"%s,%d: putback kev->time=%ld\n", __FILE__, __LINE__, kev->time);
+ 
      XPutBackEvent(im->core.display, &ev);
  
      _XimRespSyncReply(ic, buf_s[0]);
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imExten.c~
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imExten.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imExten.o
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imImSw.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imImSw.o
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imInsClbk.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imInsClbk.o
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imInt.lo
Only in libX11-1.5.0-CI-mods/modules/im/ximcp: imInt.o
diff -cibwr libX11-1.5.0/modules/im/ximcp/imLcFlt.c libX11-1.5.0-CI-mods/modules/im/ximcp/imLcFlt.c
*** libX11-1.5.0/modules/im/ximcp/imLcFlt.c	Sat Jun  2 15:37:18 2012
--- libX11-1.5.0-CI-mods/modules/im/ximcp/imLcFlt.c	Thu Oct 11 18:22:00 2012
***************
*** 120,125 ****
--- 120,131 ----
  	    /* return back to client KeyPressEvent keycode == 0 */
  	    ev->xkey.keycode = 0;
  	    ev->xkey.type = KeyPress;
+ 
+ 	    /* FIXME */
+ 	    if( KeyPress <= ev->type && ev->type <= ButtonRelease) {
+ 	      fprintf(stderr,"%s,%d: putback KeyPress ev->xkey.time=%d\n", __FILE__, __LINE__, ev->xkey.time);
+ 	    }
+ 
  	    XPutBackEvent(d, ev);
  	    if(prevcode){
  		/* For modifier key releases, restore the event, as we do not */
***************
*** 150,155 ****
--- 156,166 ----
  	ic->private.local.brl_committed = ic->private.local.brl_committing;
  	ic->private.local.composed = 0;
  	ev->xkey.keycode = 0;
+ 
+ 	if( KeyPress <= ev->type && ev->type <= ButtonRelease) {
+ 	  fprintf(stderr,"%s,%d: putback ev->xkey.time=%d\n", __FILE__, __LINE__, ev->xkey.time);
+ 	}
+ 
  	_XPutBackEvent(d, ev);
  	return(True);
      }
diff -cibwr libX11-1.5.0/modules/im/ximcp/imThaiFlt.c libX11-1.5.0-CI-mods/modules/im/ximcp/imThaiFlt.c
*** libX11-1.5.0/modules/im/ximcp/imThaiFlt.c	Sat Jun  2 15:37:18 2012
--- libX11-1.5.0-CI-mods/modules/im/ximcp/imThaiFlt.c	Thu Oct 11 18:22:12 2012
***************
*** 1416,1421 ****
--- 1416,1426 ----
      IC_SavePreviousChar(ic, new_char);
  
      ev->xkey.keycode = 0;
+ 
+     if( KeyPress <= ev->type && ev->type <= ButtonRelease) {
+       fprintf(stderr,"%s,%d: putback ev->xkey.time=%d\n", __FILE__, __LINE__, ev->xkey.time);
+     }
+ 
      XPutBackEvent(d, ev);
      return True;
  }
diff -cibwr libX11-1.5.0/modules/im/ximcp/imTrans.c libX11-1.5.0-CI-mods/modules/im/ximcp/imTrans.c
*** libX11-1.5.0/modules/im/ximcp/imTrans.c	Sat Jun  2 15:37:18 2012
--- libX11-1.5.0-CI-mods/modules/im/ximcp/imTrans.c	Thu Oct 11 18:35:43 2012
***************
*** 228,233 ****
--- 228,236 ----
      kev->display = im->core.display;
      kev->window = spec->window;
      kev->keycode = 0;
+     /* FIXME */
+     fprintf(stderr,"%s,%d: putback kev->time=%ld\n", __FILE__, __LINE__, kev->time);
+ 
      XPutBackEvent(im->core.display, &ev);
      XFlush(im->core.display);
      spec->is_putback = True;
diff -cibwr libX11-1.5.0/src/OpenDis.c libX11-1.5.0-CI-mods/src/OpenDis.c
*** libX11-1.5.0/src/OpenDis.c	Sat Jun  2 15:37:19 2012
--- libX11-1.5.0-CI-mods/src/OpenDis.c	Tue Oct 16 15:01:08 2012
***************
*** 234,239 ****
--- 234,242 ----
  	if (conn_buf_size < XLIBMINBUFSIZE)
  	    conn_buf_size = XLIBMINBUFSIZE;
  
+ 
+ 	fprintf(stderr,"OpenDis.c conn_buf_size =%ld\n", conn_buf_size);
+ 
  	if ((dpy->bufptr = dpy->buffer = Xcalloc(1, conn_buf_size)) == NULL) {
  	    OutOfMemory (dpy);
  	    return(NULL);
diff -cibwr libX11-1.5.0/src/PutBEvent.c libX11-1.5.0-CI-mods/src/PutBEvent.c
*** libX11-1.5.0/src/PutBEvent.c	Sat Jun  2 15:37:19 2012
--- libX11-1.5.0-CI-mods/src/PutBEvent.c	Tue Oct 16 17:32:29 2012
***************
*** 32,37 ****
--- 32,42 ----
  #endif
  #include "Xlibint.h"
  
+ /* FIXME */
+ #include <sys/types.h>
+ #include <unistd.h>
+ #include <stdio.h>
+ 
  int
  _XPutBackEvent (
      register Display *dpy,
***************
*** 81,86 ****
--- 86,117 ----
  	int ret;
  
  	LockDisplay(dpy);
+ 	/* FIXME */	
+ 	if(event->type == KeyPress || event->type == KeyRelease)
+ 	  {
+ 	    static FILE *fp = NULL;
+ 	    static char fname[256];
+ 	    XKeyEvent *kev = (XKeyEvent *)event;
+ 
+ #define LOGFILE "/var/tmp/X11-bug.log"
+ 
+ 	    if(fp == NULL) {
+ 	      /* snprintf(fname, 256, "/var/tmp/X11-bug.%d.log", getpid()); */
+ 	      snprintf(fname, 256, LOGFILE);
+ 	      fp = fopen(fname, "a+");
+ 	      if(fp == NULL) {
+ 		perror(fname);
+ 		exit(1);
+ 	      }
+ 	    }
+ 	      
+ 	    fprintf(fp,"XPBE: pid=%d %s time=%ld eventw=0x%08lx rootw=0x%08lx st=0x%08x serial=%lu keycode=0x%08x send_event=%d\n", 
+ 		    getpid(), event->type == KeyPress ? "press  " : "release",
+ 		    kev->time, 
+ 		    kev->window, kev->root, kev->state, kev->serial, kev->keycode, kev->send_event);
+ 	    fflush(fp);
+ 	  }
+ 
  	ret = _XPutBackEvent(dpy, event);
  	UnlockDisplay(dpy);
  	return ret;
Only in libX11-1.5.0-CI-mods/src: PutBEvent.c~

