diff --git a/pango/pango-layout.c b/pango/pango-layout.c
index b1f6346..bcec135 100644
--- a/pango/pango-layout.c
+++ b/pango/pango-layout.c
@@ -1075,10 +1075,6 @@ pango_layout_set_text (PangoLayout *layout,
     start = end;
   }
 
-  if (start != layout->text)
-    /* TODO: Write out the beginning excerpt of text? */
-    g_warning ("Invalid UTF-8 string passed to pango_layout_set_text()");
-
   layout->n_chars = pango_utf8_strlen (layout->text, -1);
 
   layout_changed (layout);
