From ab135a1c4e21553e0de185b2a3c76aec9bebb13e Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Fri, 22 Feb 2013 11:24:08 +0800
Subject: [PATCH] clean up autostart script

---
 data/autostart/openbox-autostart.in | 16 ----------------
 1 file changed, 16 deletions(-)

diff --git a/data/autostart/openbox-autostart.in b/data/autostart/openbox-autostart.in
index 5c72777..5832eb7 100755
--- a/data/autostart/openbox-autostart.in
+++ b/data/autostart/openbox-autostart.in
@@ -1,16 +1,5 @@
 #!/bin/sh
 
-# Set a background color
-BG=""
-if which hsetroot >/dev/null 2>/dev/null; then
-  BG=hsetroot
-elif which esetroot >/dev/null 2>/dev/null; then
-  BG=esetroot
-elif which xsetroot >/dev/null 2>/dev/null; then
-  BG=xsetroot
-fi
-test -z $BG || $BG -solid "#303030"
-
 GLOBALAUTOSTART="@rcdir@/autostart"
 AUTOSTART="${XDG_CONFIG_HOME:-"$HOME/.config"}/openbox/autostart"
 
@@ -27,8 +16,3 @@ if test -f $AUTOSTART; then
 elif test -f $AUTOSTART.sh; then
     sh $AUTOSTART.sh
 fi
-
-# Run the XDG autostart stuff.  These are found in /etc/xdg/autostart and
-# in $HOME/.config/autostart.  This requires PyXDG to be installed.
-# See openbox-xdg-autostart --help for more details.
-@libexecdir@/openbox-xdg-autostart "$@"
-- 
1.8.1.2

