--- a/config/baseconfig.mk	2014-04-29 08:52:58.000000000 +0800
+++ b/config/baseconfig.mk	2014-05-01 03:25:25.861373117 +0800
@@ -2,10 +2,10 @@
 # directly in python/mozbuild/mozbuild/base.py for gmake validation.
 # We thus use INCLUDED_AUTOCONF_MK to enable/disable some parts depending
 # whether a normal build is happening or whether the check is running.
-includedir := $(includedir)/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
-idldir = $(datadir)/idl/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
-installdir = $(libdir)/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
-sdkdir = $(libdir)/$(MOZ_APP_NAME)-devel-$(MOZ_APP_VERSION)
+includedir := $(includedir)/firefox
+idldir = $(includedir)/idl
+installdir = $(libdir)/firefox
+sdkdir = $(libdir)/firefox
 ifndef TOP_DIST
 TOP_DIST = dist
 endif
