--- a/xpcom/build/nsXPComInit.cpp	2014-03-21 07:07:47.000000000 +0800
+++ b/xpcom/build/nsXPComInit.cpp	2014-05-02 01:49:02.956193121 +0800
@@ -129,7 +129,7 @@
 #endif
 
 #include "ogg/ogg.h"
-#ifdef MOZ_VPX
+#if defined(MOZ_VPX) && !defined(MOZ_VPX_NO_MEM_REPORTING)
 #include "vpx_mem/vpx_mem.h"
 #endif
 
@@ -704,7 +704,7 @@
                           OggReporter::Realloc,
                           OggReporter::Free);
 
-#ifdef MOZ_VPX
+#if defined(MOZ_VPX) && !defined(MOZ_VPX_NO_MEM_REPORTING)
     // And for VPX.
     vpx_mem_set_functions(VPXReporter::Alloc,
                           VPXReporter::Calloc,
