diff -urN a/usr/lib/byobu/include/constants b/usr/lib/byobu/include/constants
--- a/usr/lib/byobu/include/constants	2013-07-27 05:02:26.000000000 +0800
+++ b/usr/lib/byobu/include/constants	2013-08-16 15:23:34.040962601 +0800
@@ -26,14 +26,14 @@
 	#ICON_F="℉"
 	ICON_C="C"
 	ICON_F="F"
-	ICON_RD="◂"
-	ICON_WR="▸"
-	ICON_MAIL="⍌"
-	ICON_UP="▴"
-	ICON_DN="▾"
-	ICON_TRASH="♸"
+	ICON_RD=""
+	ICON_WR=""
+	ICON_MAIL=""
+	ICON_UP=""
+	ICON_DN=""
+	ICON_TRASH=""
 	ICON_WIFI=
-	ICON_REBOOT="⟳"
+	ICON_REBOOT=""
 	ICON_UPDATES="!"
 	ICON_SECURITY="‼"
 	#ICON_KB="㎅"
