diff -urN a/src/VBox/Frontends/VBoxSDL/Framebuffer.cpp b/src/VBox/Frontends/VBoxSDL/Framebuffer.cpp
--- a/src/VBox/Frontends/VBoxSDL/Framebuffer.cpp	2013-03-05 22:27:09.000000000 +0800
+++ b/src/VBox/Frontends/VBoxSDL/Framebuffer.cpp	2013-03-25 04:10:22.260859088 +0800
@@ -40,7 +40,6 @@
 
 #include "VBoxSDL.h"
 #include "Framebuffer.h"
-#include "Ico64x01.h"
 
 #if defined(RT_OS_WINDOWS) || defined(RT_OS_LINUX)
 #include <SDL_syswm.h>           /* for SDL_GetWMInfo() */
@@ -67,7 +66,6 @@
 #endif /* VBOX_SECURELABEL */
 
 static bool gfSdlInitialized = false;           /**< if SDL was initialized */
-static SDL_Surface *gWMIcon = NULL;             /**< the application icon */
 static RTNATIVETHREAD gSdlNativeThread = NIL_RTNATIVETHREAD; /**< the SDL thread */
 
 //
@@ -220,17 +218,6 @@
                          RTEnvGet("SDL_VIDEODRIVER"));
     }
 
-    if (12320 == g_cbIco64x01)
-    {
-        gWMIcon = SDL_AllocSurface(SDL_SWSURFACE, 64, 64, 24, 0xff, 0xff00, 0xff0000, 0);
-        /** @todo make it as simple as possible. No PNM interpreter here... */
-        if (gWMIcon)
-        {
-            memcpy(gWMIcon->pixels, g_abIco64x01+32, g_cbIco64x01-32);
-            SDL_WM_SetIcon(gWMIcon, NULL);
-        }
-    }
-
     return true;
 }
 
@@ -245,11 +232,6 @@
     {
         AssertMsg(gSdlNativeThread == RTThreadNativeSelf(), ("Wrong thread! SDL is not threadsafe!\n"));
         SDL_QuitSubSystem(SDL_INIT_VIDEO);
-        if (gWMIcon)
-        {
-            SDL_FreeSurface(gWMIcon);
-            gWMIcon = NULL;
-        }
     }
 }
 
diff -urN a/src/VBox/Frontends/VBoxSDL/Makefile.kmk b/src/VBox/Frontends/VBoxSDL/Makefile.kmk
--- a/src/VBox/Frontends/VBoxSDL/Makefile.kmk	2013-03-05 22:27:09.000000000 +0800
+++ b/src/VBox/Frontends/VBoxSDL/Makefile.kmk	2013-03-25 04:11:27.875855399 +0800
@@ -94,15 +94,6 @@
 VBoxSDL_LDFLAGS.darwin = \
 	-framework Foundation -framework AppKit
 
-VBoxSDL_CLEAN = $(VBoxSDL_0_OUTDIR)/Ico64x01.h
-VBoxSDL_INTERMEDIATES = $(VBoxSDL_0_OUTDIR)/Ico64x01.h
-
-
-# Convert the pnm-file to a byte array.
-$$(VBoxSDL_0_OUTDIR)/Ico64x01.h: $(PATH_ROOT)/src/VBox/Frontends/VBoxSDL/ico64x01.pnm $(VBOX_BIN2C) | $$(dir $$@)
-	$(call MSG_TOOL,bin2c,VBoxSDL,$<,$@)
-	$(QUIET)$(VBOX_BIN2C) Ico64x01 $< $@
-
 # Icon include file.
 $$(VBoxSDL_0_OUTDIR)/VBoxSDL-icon.rc: $(VBOX_WINDOWS_ICON_FILE) $$(VBoxSDL_DEFPATH)/Makefile.kmk | $$(dir $$@)
 	$(RM) -f $@
