From 3fac275e6b924e4e0494748b966a20dd99b1f590 Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Mon, 6 Jan 2014 03:37:51 +0800
Subject: [PATCH] remove icon

---
 video/out/x11_common.c | 51 --------------------------------------------------
 1 file changed, 51 deletions(-)

diff --git a/video/out/x11_common.c b/video/out/x11_common.c
index 2c25c19..7e894d8 100644
--- a/video/out/x11_common.c
+++ b/video/out/x11_common.c
@@ -135,10 +135,6 @@ typedef struct
     long state;
 } MotifWmHints;
 
-static const char x11_icon[] =
-#include "video/out/x11_icon.inc"
-;
-
 static struct mp_log *x11_error_output;
 
 static void vo_x11_update_geometry(struct vo *vo);
@@ -404,7 +400,6 @@ static void init_atoms(struct vo_x11_state *x11)
     XA_INIT(_NET_WM_PID);
     XA_INIT(_NET_WM_NAME);
     XA_INIT(_NET_WM_ICON_NAME);
-    XA_INIT(_NET_WM_ICON);
     XA_INIT(_WIN_PROTOCOLS);
     XA_INIT(_WIN_LAYER);
     XA_INIT(_WIN_HINTS);
@@ -1178,51 +1173,6 @@ static bstr decompress_gz(bstr in)
 }
 #endif
 
-#define MAX_ICONS 10
-
-static void vo_x11_set_wm_icon(struct vo_x11_state *x11)
-{
-    int num_icons = 0;
-    void *icon_data[MAX_ICONS];
-    int icon_w[MAX_ICONS], icon_h[MAX_ICONS];
-
-    bstr uncompressed = decompress_gz((bstr){(char *)x11_icon, sizeof(x11_icon)});
-    bstr data = uncompressed;
-    while (data.len && num_icons < MAX_ICONS) {
-        bstr line = bstr_getline(data, &data);
-        if (bstr_eatstart0(&line, "icon: ")) {
-            int w, h;
-            if (bstr_sscanf(line, "%d %d", &w, &h) == 2) {
-                int size = w * h * 4;
-                icon_w[num_icons] = w;
-                icon_h[num_icons] = h;
-                icon_data[num_icons] = data.start;
-                num_icons++;
-                data = bstr_cut(data, size);
-            }
-        }
-    }
-
-    int icon_size = 0;
-    for (int n = 0; n < num_icons; n++)
-        icon_size += 2 + icon_w[n] * icon_h[n];
-    long *icon = talloc_array(NULL, long, icon_size);
-    long *cur = icon;
-    for (int n = 0; n < num_icons; n++) {
-        *cur++ = icon_w[n];
-        *cur++ = icon_h[n];
-        uint32_t *src = icon_data[n];
-        for (int i = 0; i < icon_h[n] * icon_w[n]; i++)
-            *cur++ = src[i];
-    }
-
-    XChangeProperty(x11->display, x11->window, x11->XA_NET_WM_ICON,
-                    XA_CARDINAL, 32, PropModeReplace,
-                    (unsigned char *)icon, icon_size);
-    talloc_free(icon);
-    talloc_free(uncompressed.start);
-}
-
 static void find_default_visual(struct vo_x11_state *x11, XVisualInfo *vis)
 {
     Display *display = x11->display;
@@ -1278,7 +1228,6 @@ static void vo_x11_create_window(struct vo *vo, XVisualInfo *vis, int x, int y,
                              NULL);
     }
 
-    vo_x11_set_wm_icon(x11);
     vo_x11_update_window_title(vo);
     vo_x11_dnd_init_window(vo);
 }
-- 
1.8.5.2

