From 4de7d5d38296f8dc3be09564a0c4609ffebf424b Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Tue, 28 Oct 2014 00:27:08 +0800
Subject: [PATCH] remove icon

---
 video/out/x11_common.c | 52 --------------------------------------------------
 1 file changed, 52 deletions(-)

diff --git a/video/out/x11_common.c b/video/out/x11_common.c
index 04f475e..3c5faa2 100644
--- a/video/out/x11_common.c
+++ b/video/out/x11_common.c
@@ -112,10 +112,6 @@ typedef struct
     long state;
 } MotifWmHints;
 
-static const char x11_icon[] =
-#include "video/out/x11_icon.inc"
-;
-
 static struct mp_log *x11_error_output;
 
 static void vo_x11_update_geometry(struct vo *vo);
@@ -1072,9 +1068,7 @@ static void vo_x11_update_window_title(struct vo *vo)
 
     const char *title = vo_get_window_title(vo);
     vo_x11_set_property_string(vo, XA_WM_NAME, title);
-    vo_x11_set_property_string(vo, XA_WM_ICON_NAME, title);
     vo_x11_set_property_utf8(vo, XA(x11, _NET_WM_NAME), title);
-    vo_x11_set_property_utf8(vo, XA(x11, _NET_WM_ICON_NAME), title);
 }
 
 #if HAVE_ZLIB
@@ -1127,51 +1121,6 @@ static bstr decompress_gz(bstr in)
 }
 #endif
 
-#define MAX_ICONS 10
-
-static void vo_x11_set_wm_icon(struct vo_x11_state *x11)
-{
-    int num_icons = 0;
-    void *icon_data[MAX_ICONS];
-    int icon_w[MAX_ICONS], icon_h[MAX_ICONS];
-
-    bstr uncompressed = decompress_gz((bstr){(char *)x11_icon, sizeof(x11_icon)});
-    bstr data = uncompressed;
-    while (data.len && num_icons < MAX_ICONS) {
-        bstr line = bstr_getline(data, &data);
-        if (bstr_eatstart0(&line, "icon: ")) {
-            int w, h;
-            if (bstr_sscanf(line, "%d %d", &w, &h) == 2) {
-                int size = w * h * 4;
-                icon_w[num_icons] = w;
-                icon_h[num_icons] = h;
-                icon_data[num_icons] = data.start;
-                num_icons++;
-                data = bstr_cut(data, size);
-            }
-        }
-    }
-
-    int icon_size = 0;
-    for (int n = 0; n < num_icons; n++)
-        icon_size += 2 + icon_w[n] * icon_h[n];
-    long *icon = talloc_array(NULL, long, icon_size);
-    long *cur = icon;
-    for (int n = 0; n < num_icons; n++) {
-        *cur++ = icon_w[n];
-        *cur++ = icon_h[n];
-        uint8_t *s = icon_data[n];
-        for (int i = 0; i < icon_h[n] * icon_w[n]; i++, s += 4)
-            *cur++ = s[0] | (s[1] << 8) | (s[2] << 16) | ((unsigned)s[3] << 24);
-    }
-
-    XChangeProperty(x11->display, x11->window, XA(x11, _NET_WM_ICON),
-                    XA_CARDINAL, 32, PropModeReplace,
-                    (unsigned char *)icon, icon_size);
-    talloc_free(icon);
-    talloc_free(uncompressed.start);
-}
-
 static void vo_x11_create_window(struct vo *vo, XVisualInfo *vis,
                                  struct mp_rect rc)
 {
@@ -1223,7 +1172,6 @@ static void vo_x11_create_window(struct vo *vo, XVisualInfo *vis,
     }
 
     if (vo->opts->WinID < 0) {
-        vo_x11_set_wm_icon(x11);
         vo_x11_update_window_title(vo);
         vo_x11_dnd_init_window(vo);
     }
-- 
2.1.2

