From 3e84ce60a27c0558309f5cfdd60cef46cee4fbd1 Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Fri, 20 Dec 2013 15:44:57 +0800
Subject: [PATCH] remove icon

---
 video/out/x11_common.c | 105 -------------------------------------------------
 1 file changed, 105 deletions(-)

diff --git a/video/out/x11_common.c b/video/out/x11_common.c
index a042bd0..11e0e65 100644
--- a/video/out/x11_common.c
+++ b/video/out/x11_common.c
@@ -66,10 +66,6 @@
 #include <X11/XF86keysym.h>
 #endif
 
-#if HAVE_ZLIB
-#include <zlib.h>
-#endif
-
 #include "input/input.h"
 #include "input/keycodes.h"
 
@@ -132,10 +128,6 @@ typedef struct
     long state;
 } MotifWmHints;
 
-static const char x11_icon[] =
-#include "video/out/x11_icon.inc"
-;
-
 static void vo_x11_update_geometry(struct vo *vo);
 static void vo_x11_fullscreen(struct vo *vo);
 static int vo_x11_get_fs_type(struct vo *vo);
@@ -402,7 +394,6 @@ static void init_atoms(struct vo_x11_state *x11)
     XA_INIT(_NET_WM_PID);
     XA_INIT(_NET_WM_NAME);
     XA_INIT(_NET_WM_ICON_NAME);
-    XA_INIT(_NET_WM_ICON);
     XA_INIT(_WIN_PROTOCOLS);
     XA_INIT(_WIN_LAYER);
     XA_INIT(_WIN_HINTS);
@@ -974,101 +965,6 @@ static void vo_x11_update_window_title(struct vo *vo)
     vo_x11_set_property_utf8(vo, x11->XA_NET_WM_ICON_NAME, title);
 }
 
-#if HAVE_ZLIB
-static bstr decompress_gz(bstr in)
-{
-    bstr res = {0};
-    z_stream zstream;
-    uint8_t *dest;
-    size_t size = in.len;
-    int result;
-
-    zstream.zalloc = (alloc_func) 0;
-    zstream.zfree = (free_func) 0;
-    zstream.opaque = (voidpf) 0;
-    // 32 for gzip header, 15 for max. window bits
-    if (inflateInit2(&zstream, 32 + 15) != Z_OK)
-        goto error;
-    zstream.next_in = (Bytef *) in.start;
-    zstream.avail_in = size;
-
-    dest = NULL;
-    zstream.avail_out = size;
-    do {
-        size += 4000;
-        dest = talloc_realloc_size(NULL, dest, size);
-        zstream.next_out = (Bytef *) (dest + zstream.total_out);
-        result = inflate(&zstream, Z_NO_FLUSH);
-        if (result != Z_OK && result != Z_STREAM_END) {
-            talloc_free(dest);
-            dest = NULL;
-            inflateEnd(&zstream);
-            goto error;
-        }
-        zstream.avail_out += 4000;
-    } while (zstream.avail_out == 4000 && zstream.avail_in != 0
-             && result != Z_STREAM_END);
-
-    size = zstream.total_out;
-    inflateEnd(&zstream);
-
-    res.start = dest;
-    res.len = size;
-error:
-    return res;
-}
-#else
-static bstr decompress_gz(bstr in)
-{
-    return (bstr){0};
-}
-#endif
-
-#define MAX_ICONS 10
-
-static void vo_x11_set_wm_icon(struct vo_x11_state *x11)
-{
-    int num_icons = 0;
-    void *icon_data[MAX_ICONS];
-    int icon_w[MAX_ICONS], icon_h[MAX_ICONS];
-
-    bstr uncompressed = decompress_gz((bstr){(char *)x11_icon, sizeof(x11_icon)});
-    bstr data = uncompressed;
-    while (data.len && num_icons < MAX_ICONS) {
-        bstr line = bstr_getline(data, &data);
-        if (bstr_eatstart0(&line, "icon: ")) {
-            int w, h;
-            if (bstr_sscanf(line, "%d %d", &w, &h) == 2) {
-                int size = w * h * 4;
-                icon_w[num_icons] = w;
-                icon_h[num_icons] = h;
-                icon_data[num_icons] = data.start;
-                num_icons++;
-                data = bstr_cut(data, size);
-            }
-        }
-    }
-
-    int icon_size = 0;
-    for (int n = 0; n < num_icons; n++)
-        icon_size += 2 + icon_w[n] * icon_h[n];
-    long *icon = talloc_array(NULL, long, icon_size);
-    long *cur = icon;
-    for (int n = 0; n < num_icons; n++) {
-        *cur++ = icon_w[n];
-        *cur++ = icon_h[n];
-        uint32_t *src = icon_data[n];
-        for (int i = 0; i < icon_h[n] * icon_w[n]; i++)
-            *cur++ = src[i];
-    }
-
-    XChangeProperty(x11->display, x11->window, x11->XA_NET_WM_ICON,
-                    XA_CARDINAL, 32, PropModeReplace,
-                    (unsigned char *)icon, icon_size);
-    talloc_free(icon);
-    talloc_free(uncompressed.start);
-}
-
 static void find_default_visual(struct vo_x11_state *x11, XVisualInfo *vis)
 {
     Display *display = x11->display;
@@ -1124,7 +1020,6 @@ static void vo_x11_create_window(struct vo *vo, XVisualInfo *vis, int x, int y,
                              NULL);
     }
 
-    vo_x11_set_wm_icon(x11);
     vo_x11_update_window_title(vo);
 }
 
-- 
1.8.5.2

