From afe4c6d08c60e736324a31c62bac7746ab98b800 Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Sun, 8 Sep 2013 17:13:30 +0800
Subject: [PATCH 2/2] mute no bind found for key

---
 mpvcore/input/input.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/mpvcore/input/input.c b/mpvcore/input/input.c
index bf073d4..edade11 100644
--- a/mpvcore/input/input.c
+++ b/mpvcore/input/input.c
@@ -899,11 +899,6 @@ static int parse_cmd(struct mp_cmd **dest, bstr str, const char *loc)
         bstr old = bstr0(entry->old);
         bool silent = bstr_eatstart0(&old, "!");
         if (bstrcasecmp(bstr_splice(str, 0, old.len), old) == 0) {
-            if (!silent) {
-                mp_tmsg(MSGT_INPUT, MSGL_WARN, "Warning: command '%.*s' is "
-                        "deprecated, replaced with '%s' at %s.\n",
-                        BSTR_P(old), entry->new, loc);
-            }
             bstr s = bstr_cut(str, old.len);
             str = bstr0(talloc_asprintf(tmp, "%s%.*s", entry->new, BSTR_P(s)));
             start = str;
-- 
1.8.3.2

