From 0ba50ab118e867180c31680a5c94f5a0629e6256 Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Wed, 7 Aug 2013 06:00:06 +0800
Subject: [PATCH] mute no bind found for key

---
 mpvcore/input/input.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/mpvcore/input/input.c b/mpvcore/input/input.c
index 935ed62..9a85c92 100644
--- a/mpvcore/input/input.c
+++ b/mpvcore/input/input.c
@@ -1294,8 +1294,6 @@ static mp_cmd_t *get_cmd_from_keys(struct input_ctx *ictx, char *force_section,
 
     if (cmd == NULL) {
         char *key_buf = get_key_combo_name(keys, n);
-        mp_tmsg(MSGT_INPUT, MSGL_WARN,
-                "No bind found for key '%s'.\n", key_buf);
         talloc_free(key_buf);
         return NULL;
     }
-- 
1.8.3.2

