From 329a3f405b4d37db8af86fff3c8fdd2b53f37e56 Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Wed, 16 Apr 2014 02:37:31 +0800
Subject: [PATCH] remove INSTALL_MASK info

---
 bin/misc-functions.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/misc-functions.sh b/bin/misc-functions.sh
index 5ccf7c2..61be9d3 100755
--- a/bin/misc-functions.sh
+++ b/bin/misc-functions.sh
@@ -982,7 +982,7 @@ install_mask() {
 	local no_inst
 	for no_inst in ${install_mask}; do
 		set +o noglob
-		__quiet_mode || einfo "Removing ${no_inst}"
+		__quiet_mode
 		# normal stuff
 		rm -Rf "${root}"/${no_inst} >&/dev/null
 
-- 
1.9.2

