diff --git a/gui/Gtk2Gui.c b/gui/Gtk2Gui.c
index ac8eabc..30f84cb 100755
--- a/gui/Gtk2Gui.c
+++ b/gui/Gtk2Gui.c
@@ -236,8 +236,6 @@ void StartGui() {
 
 	Window = glade_xml_get_widget(xml, "MainWindow");
 	gtk_window_set_title(GTK_WINDOW(Window), "PCSXR");
-	gtk_window_set_icon_from_file(GTK_WINDOW(Window), PIXMAPDIR "pcsxr-icon.png", NULL);
-	gtk_window_set_default_icon_from_file(PIXMAPDIR "pcsxr-icon.png", NULL);
 	ResetMenuSlots(xml);
 
 	// Set up callbacks