diff --git a/gui/Gtk2Gui.c b/gui/Gtk2Gui.c
index a0d3b3f..6ae98d6 100755
--- a/gui/Gtk2Gui.c
+++ b/gui/Gtk2Gui.c
@@ -248,8 +248,6 @@ void StartGui() {
 	gtk_widget_show(GTK_WIDGET(Window));
 
 	gtk_window_set_title(GTK_WINDOW(Window), "PCSXR");
-	gtk_window_set_icon_from_file(GTK_WINDOW(Window), PIXMAPDIR "pcsxr-icon.png", NULL);
-	gtk_window_set_default_icon_from_file(PIXMAPDIR "pcsxr-icon.png", NULL);
 	ResetMenuSlots();
 
 	// Set up callbacks