From 13a0ed3c7da314509a11f9240c8a780e6636f37e Mon Sep 17 00:00:00 2001
From: stlifey joe <stlifey@gmail.com>
Date: Thu, 14 Feb 2013 10:55:44 +0800
Subject: [PATCH] mute no bind found for key

---
 core/input/input.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/core/input/input.c b/core/input/input.c
index c12dba2..50bd011 100644
--- a/core/input/input.c
+++ b/core/input/input.c
@@ -1224,8 +1224,6 @@ static mp_cmd_t *get_cmd_from_keys(struct input_ctx *ictx, int n, int *keys)
 
     if (cmd == NULL) {
         char *key_buf = get_key_combo_name(keys, n);
-        mp_tmsg(MSGT_INPUT, MSGL_WARN,
-                "No bind found for key '%s'.\n", key_buf);
         talloc_free(key_buf);
         return NULL;
     }
-- 
1.8.1.2

